Xamarin.Essentials/Types/FileProvider.android.cs

```diff
@@ -0,0 +1,102 @@
+
+        public static FileProviderLocation TemporaryLocation { get; set; }
+
+        internal static Java.IO.File GetTemporaryDirectory()
+        {
+            var root = GetTemporaryRootDirectory();
+            var dir = new Java.IO.File(root, "2203693cc04e0be7f4f024d5f9499e13");
+            dir.Mkdirs();
+            dir.DeleteOnExit();
+            return dir;
+        }
+
+        internal static Java.IO.File GetTemporaryRootDirectory()
+        {
+            // we want the internal storage
+            if (TemporaryLocation == FileProviderLocation.Internal)
+                return Platform.AppContext.CacheDir;
+
+            // if we need permission, then check
+            // if we can fall back, then don't throw
+            var externalOnly = TemporaryLocation == FileProviderLocation.External;
+            var hasPermission = Platform.HasApiLevel(BuildVersionCodes.Kitkat) || Permissions.EnsureDeclared(PermissionType.ExternalStorage, externalOnly);
+
+            // make sure the external storage is available
+            var hasExternalMedia = Platform.HasApiLevel(BuildVersionCodes.Lollipop)
+                ? AndroidEnvironment.GetExternalStorageState(Platform.AppContext.ExternalCacheDir) == AndroidEnvironment.MediaMounted
+#pragma warning disable CS0618 // Type or member is obsolete
+                : AndroidEnvironment.GetStorageState(Platform.AppContext.ExternalCacheDir) == AndroidEnvironment.MediaMounted;
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            // undo all the work if we have requested a fail (mainly for testing)
+            if (AlwaysFailExternalMediaAccess)
+                hasExternalMedia = false;
+
+            // fail if we need the external storage, but there is none
+            if (externalOnly && !hasExternalMedia)
+                throw new InvalidOperationException("Unable to access the external storage, the media is not mounted.");
+
+            // based on permssions, return the correct directory
+            // if permission were required, then it would have already thrown
+            return hasPermission && hasExternalMedia
+                ? Platform.AppContext.ExternalCacheDir
+                : Platform.AppContext.CacheDir;
+        }
+
+        internal static bool IsFileInPublicLocation(string filename)
+        {
+            // get the Android path, we use "CanonicalPath" instead of "AbsolutePath"
+            // because we want to resolve any ".." and links/redirects
+            var file = new Java.IO.File(filename);
+            filename = file.CanonicalPath;
+
+            // the shared paths from the "xamarin_essentials_fileprovider_file_paths.xml" resource
+            var publicLocations = new[]
+            {
+                AndroidEnvironment.ExternalStorageDirectory.CanonicalPath,
+                Platform.AppContext.ExternalCacheDir.CanonicalPath,
+                Platform.AppContext.CacheDir.CanonicalPath,
+            };
```

Comment:
These are the locations we exposed in our xml resource, we _may_ want to look at the resource file and dunamically create it. But, for now I think this is just fine.